controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -419,52 +419,74 @@ func (r *BareMetalHostReconciler) actionUnmanaged(prov provisioner.Provisioner,
 }
 
 // Test the credentials by connecting to the management controller.
 func (r *BareMetalHostReconciler) actionRegistering(prov provisioner.Provisioner, info *reconcileInfo) actionResult {
 	info.log.Info("registering and validating access to management controller",
 		"credentials", info.host.Status.TriedCredentials)
+	dirty := false
 
 	credsChanged := !info.host.Status.TriedCredentials.Match(*info.bmcCredsSecret)
 	if credsChanged {
 		info.log.Info("new credentials")
 		info.host.UpdateTriedCredentials(*info.bmcCredsSecret)
 		info.postSaveCallbacks = append(info.postSaveCallbacks, updatedCredentials.Inc)
+		dirty = true
 	}
 
-	provResult, err := prov.ValidateManagementAccess(credsChanged, info.host.Status.ErrorType == metal3v1alpha1.RegistrationError)
+	provResult, provID, err := prov.ValidateManagementAccess(credsChanged, info.host.Status.ErrorType == metal3v1alpha1.RegistrationError)
 	if err != nil {
 		noManagementAccess.Inc()
 		return actionError{errors.Wrap(err, "failed to validate BMC access")}
 	}
 
-	info.log.Info("response from validate", "provResult", provResult)
-
 	if provResult.ErrorMessage != "" {
 		return recordActionFailure(info, metal3v1alpha1.RegistrationError, provResult.ErrorMessage)
 	}
 
+	if provID != "" && info.host.Status.Provisioning.ID != provID {
+		info.log.Info("setting provisioning id", "ID", provID)
+		info.host.Status.Provisioning.ID = provID
+		dirty = true
+	}
+
 	if provResult.Dirty {
 		info.log.Info("host not ready", "wait", provResult.RequeueAfter)
-		clearError(info.host)
-		return actionContinue{provResult.RequeueAfter}
+		result := actionContinue{provResult.RequeueAfter}
+		if clearError(info.host) {
+			dirty = true
+		}
+		if dirty {
+			return actionUpdate{result}
+		}
+		return result
 	}
 
 	// Reaching this point means the credentials are valid and worked,
 	// so clear any previous error and record the success in the
 	// status block.
-	info.log.Info("updating credentials success status fields")
-	registeredNewCreds := !info.host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
-	info.host.UpdateGoodCredentials(*info.bmcCredsSecret)
-	info.log.Info("clearing previous error message")
-	clearError(info.host)
-
-	if registeredNewCreds {
+	if !info.host.Status.GoodCredentials.Match(*info.bmcCredsSecret) {
+		info.log.Info("updating credentials success status fields")
+		info.host.UpdateGoodCredentials(*info.bmcCredsSecret)
 		info.publishEvent("BMCAccessValidated", "Verified access to BMC")
+		dirty = true
+	} else {
+		info.log.Info("verified access to the BMC")
+	}
+	if clearError(info.host) {
+		info.log.Info("clearing previous error message")
+		dirty = true
 	}
 
+	if dirty {
+		// Normally returning actionComplete would be sufficient to ensure
+		// changes are written, but when this is called in a state other than
+		// Registering, the state machine will ignore actionComplete and
+		// proceed with the action function for the current state.
+		return actionUpdate{}
```

Comment:
The comment isn't clear to me, isn't the update driven by the `needsReregister` flag in case of `actionComplete` regardless of the state? Not sure if it could help to return also the `dirty` flag as part of this func

Reply:
I didn't actually look at the state machine code before writing this comment, so yeah it's more subtle.
The effect is the same though, I think: if this returns actionComplete then ensureRegistered will either return actionContinue or nil (which will proceed with the action function for the current state). In either case nothing gets written to the Status, so we have to return actionUpdate to make sure that writes happen.

Reply:
I added another commit at the end to rename the method and simplify both it and the caller, since this in no longer an 'action' method proper.